File: golang/internal/engine/handler.go
package engine

import (
	"errors"
	"fmt"

	"github.com/gofiber/fiber/v2"
	"github.com/jackc/pgx/v5/pgconn"

	"rocket-backend/internal/metadata"
	"rocket-backend/internal/store"
)

type Handler struct {
	store    *store.Store
	registry *metadata.Registry
}

func NewHandler(s *store.Store, reg *metadata.Registry) *Handler {
	return &Handler{store: s, registry: reg}
}

// List handles GET /api/:entity
func (h *Handler) List(c *fiber.Ctx) error {
	entity, err := h.resolveEntity(c)
	if err != nil {
		return err
	}

	user := getUser(c)
	if err := CheckPermission(user, entity.Name, "read", h.registry, nil); err != nil {
		return err
	}

	plan, err := ParseQueryParams(c, entity, h.registry)
	if err != nil {
		return err
	}

	// Inject row-level security filters
	if filters := GetReadFilters(user, entity.Name, h.registry); len(filters) > 0 {
		plan.Filters = append(plan.Filters, filters...)
	}

	// Execute data query
	qr := BuildSelectSQL(plan)
	rows, err := store.QueryRows(c.Context(), h.store.Pool, qr.SQL, qr.Params...)
	if err != nil {
		return fmt.Errorf("list %s: %w", entity.Name, err)
	}

	// Execute count query
	cr := BuildCountSQL(plan)
	countRow, err := store.QueryRow(c.Context(), h.store.Pool, cr.SQL, cr.Params...)
	if err != nil {
		return fmt.Errorf("count %s: %w", entity.Name, err)
	}
	total := countRow["count"]

	// Load includes
	if len(plan.Includes) > 0 {
		if err := LoadIncludes(c.Context(), h.store.Pool, h.registry, entity, rows, plan.Includes); err != nil {
			return fmt.Errorf("load includes: %w", err)
		}
	}

	// Ensure non-nil slice for JSON
	if rows == nil {
		rows = []map[string]any{}
	}

	return c.JSON(fiber.Map{
		"data": rows,
		"meta": fiber.Map{
			"page":     plan.Page,
			"per_page": plan.PerPage,
			"total":    total,
		},
	})
}

// GetByID handles GET /api/:entity/:id
func (h *Handler) GetByID(c *fiber.Ctx) error {
	entity, err := h.resolveEntity(c)
	if err != nil {
		return err
	}

	user := getUser(c)
	if err := CheckPermission(user, entity.Name, "read", h.registry, nil); err != nil {
		return err
	}

	id := c.Params("id")
	row, err := fetchRecord(c.Context(), h.store.Pool, entity, id)
	if err != nil {
		if errors.Is(err, store.ErrNotFound) {
			return respondError(c, NotFoundError(entity.Name, id))
		}
		return fmt.Errorf("get %s/%s: %w", entity.Name, id, err)
	}

	// Load includes
	includes := parseIncludes(c)
	if len(includes) > 0 {
		rows := []map[string]any{row}
		if err := LoadIncludes(c.Context(), h.store.Pool, h.registry, entity, rows, includes); err != nil {
			return fmt.Errorf("load includes: %w", err)
		}
		row = rows[0]
	}

	return c.JSON(fiber.Map{"data": row})
}

// Create handles POST /api/:entity
func (h *Handler) Create(c *fiber.Ctx) error {
	entity, err := h.resolveEntity(c)
	if err != nil {
		return err
	}

	user := getUser(c)
	if err := CheckPermission(user, entity.Name, "create", h.registry, nil); err != nil {
		return err
	}

	var body map[string]any
	if err := c.BodyParser(&body); err != nil {
		return respondError(c, NewAppError("INVALID_PAYLOAD", 400, "Invalid JSON body"))
	}

	plan, validationErrs := PlanWrite(entity, h.registry, body, nil)
	if len(validationErrs) > 0 {
		return respondError(c, ValidationError(validationErrs))
	}

	record, err := ExecuteWritePlan(c.Context(), h.store, h.registry, plan)
	if err != nil {
		return handleWriteError(c, err)
	}

	return c.Status(201).JSON(fiber.Map{"data": record})
}

// Update handles PUT /api/:entity/:id
func (h *Handler) Update(c *fiber.Ctx) error {
	entity, err := h.resolveEntity(c)
	if err != nil {
		return err
	}

	id := c.Params("id")

	// Verify record exists and check permissions against current state
	currentRecord, err := fetchRecord(c.Context(), h.store.Pool, entity, id)
	if err != nil {
		if errors.Is(err, store.ErrNotFound) {
			return respondError(c, NotFoundError(entity.Name, id))
		}
		return fmt.Errorf("fetch %s/%s: %w", entity.Name, id, err)
	}

	user := getUser(c)
	if err := CheckPermission(user, entity.Name, "update", h.registry, currentRecord); err != nil {
		return err
	}

	var body map[string]any
	if err := c.BodyParser(&body); err != nil {
		return respondError(c, NewAppError("INVALID_PAYLOAD", 400, "Invalid JSON body"))
	}

	plan, validationErrs := PlanWrite(entity, h.registry, body, id)
	if len(validationErrs) > 0 {
		return respondError(c, ValidationError(validationErrs))
	}

	record, err := ExecuteWritePlan(c.Context(), h.store, h.registry, plan)
	if err != nil {
		return handleWriteError(c, err)
	}

	return c.JSON(fiber.Map{"data": record})
}

// Delete handles DELETE /api/:entity/:id
func (h *Handler) Delete(c *fiber.Ctx) error {
	entity, err := h.resolveEntity(c)
	if err != nil {
		return err
	}

	id := c.Params("id")

	// Check permissions against current record
	currentRecord, err := fetchRecord(c.Context(), h.store.Pool, entity, id)
	if err != nil {
		if errors.Is(err, store.ErrNotFound) {
			return respondError(c, NotFoundError(entity.Name, id))
		}
		return fmt.Errorf("fetch %s/%s: %w", entity.Name, id, err)
	}

	user := getUser(c)
	if err := CheckPermission(user, entity.Name, "delete", h.registry, currentRecord); err != nil {
		return err
	}

	tx, err := h.store.BeginTx(c.Context())
	if err != nil {
		return fmt.Errorf("begin tx: %w", err)
	}
	defer tx.Rollback(c.Context()) //nolint:errcheck

	// Handle cascades
	if err := HandleCascadeDelete(c.Context(), tx, h.registry, entity, id); err != nil {
		var appErr *AppError
		if errors.As(err, &appErr) {
			return respondError(c, appErr)
		}
		return fmt.Errorf("cascade delete: %w", err)
	}

	// Delete the record
	var sql string
	var params []any
	if entity.SoftDelete {
		sql, params = BuildSoftDeleteSQL(entity, id)
	} else {
		sql, params = BuildHardDeleteSQL(entity, id)
	}

	affected, err := store.Exec(c.Context(), tx, sql, params...)
	if err != nil {
		return fmt.Errorf("delete %s/%s: %w", entity.Name, id, err)
	}
	if affected == 0 {
		return respondError(c, NotFoundError(entity.Name, id))
	}

	if err := tx.Commit(c.Context()); err != nil {
		return fmt.Errorf("commit: %w", err)
	}

	return c.JSON(fiber.Map{"data": fiber.Map{"id": id}})
}

func (h *Handler) resolveEntity(c *fiber.Ctx) (*metadata.Entity, error) {
	name := c.Params("entity")
	entity := h.registry.GetEntity(name)
	if entity == nil {
		return nil, respondError(c, UnknownEntityError(name))
	}
	return entity, nil
}

func getUser(c *fiber.Ctx) *metadata.UserContext {
	user, _ := c.Locals("user").(*metadata.UserContext)
	return user
}

func respondError(c *fiber.Ctx, appErr *AppError) error {
	return c.Status(appErr.Status).JSON(ErrorResponse{Error: appErr})
}

func handleWriteError(c *fiber.Ctx, err error) error {
	var appErr *AppError
	if errors.As(err, &appErr) {
		return respondError(c, appErr)
	}

	if errors.Is(err, store.ErrUniqueViolation) {
		msg := "A record with this value already exists"
		var pgErr *pgconn.PgError
		if errors.As(err, &pgErr) && pgErr.Detail != "" {
			msg = pgErr.Detail
		}
		return respondError(c, ConflictError(msg))
	}

	return err
}

func parseIncludes(c *fiber.Ctx) []string {
	inc := c.Query("include")
	if inc == "" {
		return nil
	}
	var includes []string
	for _, name := range splitAndTrim(inc) {
		includes = append(includes, name)
	}
	return includes
}

func splitAndTrim(s string) []string {
	parts := make([]string, 0)
	for _, p := range splitComma(s) {
		trimmed := trimSpace(p)
		if trimmed != "" {
			parts = append(parts, trimmed)
		}
	}
	return parts
}

func splitComma(s string) []string {
	var parts []string
	start := 0
	for i := 0; i < len(s); i++ {
		if s[i] == ',' {
			parts = append(parts, s[start:i])
			start = i + 1
		}
	}
	parts = append(parts, s[start:])
	return parts
}

func trimSpace(s string) string {
	start := 0
	end := len(s)
	for start < end && s[start] == ' ' {
		start++
	}
	for end > start && s[end-1] == ' ' {
		end--
	}
	return s[start:end]
}

File: expressjs/src/engine/handler.ts
import type { Request, Response, NextFunction } from "express";
import type { Store } from "../store/postgres.js";
import { queryRows, queryRow, exec, ErrNotFound } from "../store/postgres.js";
import type { Registry } from "../metadata/registry.js";
import { AppError, notFoundError, unknownEntityError } from "./errors.js";
import { parseQueryParams, buildSelectSQL, buildCountSQL } from "./query.js";
import { buildSoftDeleteSQL, buildHardDeleteSQL } from "./writer.js";
import { planWrite, executeWritePlan, fetchRecord } from "./nested-write.js";
import { loadIncludes } from "./includes.js";
import { handleCascadeDelete } from "./soft-delete.js";
import { checkPermission, getReadFilters } from "../auth/permissions.js";

type AsyncHandler = (
  req: Request,
  res: Response,
  next: NextFunction,
) => Promise<void>;

function asyncHandler(fn: AsyncHandler) {
  return (req: Request, res: Response, next: NextFunction) => {
    fn(req, res, next).catch(next);
  };
}

export class Handler {
  private store: Store;
  private registry: Registry;

  constructor(store: Store, registry: Registry) {
    this.store = store;
    this.registry = registry;
  }

  list = asyncHandler(async (req: Request, res: Response) => {
    const entity = this.resolveEntity(req);

    checkPermission(req.user, entity.name, "read", this.registry, null);

    const plan = parseQueryParams(req, entity, this.registry);

    // Inject row-level security filters
    const filters = getReadFilters(req.user, entity.name, this.registry);
    plan.filters.push(...filters);

    // Execute data query
    const qr = buildSelectSQL(plan);
    let rows = await queryRows(this.store.pool, qr.sql, qr.params);

    // Execute count query
    const cr = buildCountSQL(plan);
    const countRow = await queryRow(this.store.pool, cr.sql, cr.params);
    const total = parseInt(String(countRow.count), 10);

    // Load includes
    if (plan.includes.length > 0) {
      await loadIncludes(
        this.store.pool,
        this.registry,
        entity,
        rows,
        plan.includes,
      );
    }

    if (!rows) rows = [];

    res.json({
      data: rows,
      meta: {
        page: plan.page,
        per_page: plan.perPage,
        total,
      },
    });
  });

  getById = asyncHandler(async (req: Request, res: Response) => {
    const entity = this.resolveEntity(req);

    checkPermission(req.user, entity.name, "read", this.registry, null);

    const id = req.params.id;

    let row: Record<string, any>;
    try {
      row = await fetchRecord(this.store.pool, entity, id);
    } catch (err) {
      if (err === ErrNotFound) {
        throw notFoundError(entity.name, id);
      }
      throw err;
    }

    // Load includes
    const includes = parseIncludesParam(req);
    if (includes.length > 0) {
      const rows = [row];
      await loadIncludes(
        this.store.pool,
        this.registry,
        entity,
        rows,
        includes,
      );
      row = rows[0];
    }

    res.json({ data: row });
  });

  create = asyncHandler(async (req: Request, res: Response) => {
    const entity = this.resolveEntity(req);

    checkPermission(req.user, entity.name, "create", this.registry, null);

    const body = req.body;

    if (!body || typeof body !== "object") {
      throw new AppError("INVALID_PAYLOAD", 400, "Invalid JSON body");
    }

    const plan = planWrite(entity, this.registry, body, null);
    const record = await executeWritePlan(this.store, this.registry, plan);

    res.status(201).json({ data: record });
  });

  update = asyncHandler(async (req: Request, res: Response) => {
    const entity = this.resolveEntity(req);
    const id = req.params.id;

    // Verify record exists and check permissions against current state
    let currentRecord: Record<string, any>;
    try {
      currentRecord = await fetchRecord(this.store.pool, entity, id);
    } catch (err) {
      if (err === ErrNotFound) {
        throw notFoundError(entity.name, id);
      }
      throw err;
    }

    checkPermission(req.user, entity.name, "update", this.registry, currentRecord);

    const body = req.body;
    if (!body || typeof body !== "object") {
      throw new AppError("INVALID_PAYLOAD", 400, "Invalid JSON body");
    }

    const plan = planWrite(entity, this.registry, body, id);
    const record = await executeWritePlan(this.store, this.registry, plan);

    res.json({ data: record });
  });

  delete = asyncHandler(async (req: Request, res: Response) => {
    const entity = this.resolveEntity(req);
    const id = req.params.id;

    // Check permissions against current record
    let currentRecord: Record<string, any>;
    try {
      currentRecord = await fetchRecord(this.store.pool, entity, id);
    } catch (err) {
      if (err === ErrNotFound) {
        throw notFoundError(entity.name, id);
      }
      throw err;
    }

    checkPermission(req.user, entity.name, "delete", this.registry, currentRecord);

    const client = await this.store.beginTx();
    try {
      // Handle cascades
      await handleCascadeDelete(client, this.registry, entity, id);

      // Delete the record
      let sql: string;
      let params: any[];
      if (entity.soft_delete) {
        [sql, params] = buildSoftDeleteSQL(entity, id);
      } else {
        [sql, params] = buildHardDeleteSQL(entity, id);
      }

      const affected = await exec(client, sql, params);
      if (affected === 0) {
        throw notFoundError(entity.name, id);
      }

      await client.query("COMMIT");
      res.json({ data: { id } });
    } catch (err) {
      await client.query("ROLLBACK");
      throw err;
    } finally {
      client.release();
    }
  });

  private resolveEntity(req: Request) {
    const name = req.params.entity;
    const entity = this.registry.getEntity(name);
    if (!entity) {
      throw unknownEntityError(name);
    }
    return entity;
  }
}

function parseIncludesParam(req: Request): string[] {
  const inc = req.query.include as string | undefined;
  if (!inc) return [];
  return inc
    .split(",")
    .map((s) => s.trim())
    .filter(Boolean);
}
